File: Sources/Lib/Audio/AudioBuffers.swift
// Class-backed, contiguous per-channel audio storage.
//
// Replaces the old `[[PrcFmt]]` ("array of arrays") chunk storage. The 2-D
// nested-array layout had two costs the audio thread couldn't afford:
//
//   1. `array[ch].withUnsafeMutableBufferPointer { ... }` triggers Swift's
//      uniqueness check on the inner buffer; whenever any external reference
//      kept the inner array's storage shared (closures, queues, captures),
//      the next mutable access malloc'd a fresh copy.
//   2. The outer array's element copies bumped per-channel buffer refcounts
//      on every value-copy of an `AudioChunk`.
//
// `AudioBuffers` allocates one contiguous block of `channels * capacity`
// `PrcFmt` values up front and exposes per-channel `UnsafeMutableBufferPointer`
// views that are stable for the buffer's lifetime. The hot path uses the
// pointers directly — no `withUnsafe*` round trips, no uniqueness checks.
//
// Thread-safety: `AudioBuffers` itself does no synchronisation. The pipeline
// already enforces single-writer discipline (the audio thread owns each
// buffer while it processes a chunk), so `@unchecked Sendable` is honest
// here — the type is no less safe than the `[[PrcFmt]]` it replaces.

import Foundation

/// Contiguous, per-channel audio storage backed by a single heap allocation.
public final class AudioBuffers: @unchecked Sendable {
  /// Number of channels.
  public let channels: Int
  /// Per-channel capacity in `PrcFmt` samples.
  public let capacity: Int

  /// One contiguous `channels * capacity` block. Channel `ch` lives at
  /// `[ch * capacity ..< (ch + 1) * capacity]`.
  private let storage: UnsafeMutablePointer<PrcFmt>

  /// Pre-built per-channel views — sized to `capacity`, pointing into
  /// `storage`. Built once at init and never resized; the pointers stay
  /// valid for the entire lifetime of this `AudioBuffers`.
  @usableFromInline
  internal let channelBuffers: [MutableWaveform]

  /// Allocate a fresh buffer pool, zero-initialised.
  public init(channels: Int, capacity: Int) {
    precondition(channels > 0, "channels must be positive")
    precondition(capacity > 0, "capacity must be positive")
    self.channels = channels
    self.capacity = capacity

    let total = channels * capacity
    let storage = UnsafeMutablePointer<PrcFmt>.allocate(capacity: total)
    storage.initialize(repeating: 0, count: total)
    self.storage = storage

    var bufs: [MutableWaveform] = []
    bufs.reserveCapacity(channels)
    for ch in 0..<channels {
      bufs.append(
        MutableWaveform(start: storage + ch * capacity, count: capacity))
    }
    self.channelBuffers = bufs
  }

  deinit {
    storage.deinitialize(count: channels * capacity)
    storage.deallocate()
  }

  /// Mutable per-channel pointer. The pointer is stable for the lifetime
  /// of the `AudioBuffers`; callers may cache it.
  @inlinable
  public subscript(ch: Int) -> MutableWaveform {
    channelBuffers[ch]
  }
}

File: Sources/CDSP/Audio/audio_buffers.c
// Heap-allocated, contiguous per-channel audio storage.
//
// Replaces nested array/pointer-to-pointer ("array of arrays") chunk storage.
#include "audio_buffers.h"

#include <stdlib.h>
#include <string.h>

struct audio_buffers {
  size_t channels;
  size_t capacity;
  double* storage;
  mutable_waveform_t* channel_buffers;
};

size_t audio_buffers_get_channels(const audio_buffers_t* buffers) {
  return buffers ? buffers->channels : 0;
}

size_t audio_buffers_get_capacity(const audio_buffers_t* buffers) {
  return buffers ? buffers->capacity : 0;
}

mutable_waveform_t audio_buffers_get_channel(const audio_buffers_t* buffers,
                                             size_t ch) {
  if (!buffers || ch >= buffers->channels) return NULL;
  return buffers->channel_buffers[ch];
}

/// Allocate a fresh buffer pool, zero-initialised.
audio_buffers_t* audio_buffers_create(size_t channels, size_t capacity) {
  if (channels == 0 || capacity == 0) return NULL;

  // Check for overflow before multiplication
  if (capacity > SIZE_MAX / channels) return NULL;

  audio_buffers_t* buf = (audio_buffers_t*)malloc(sizeof(audio_buffers_t));
  if (!buf) return NULL;

  buf->channels = channels;
  buf->capacity = capacity;
  buf->storage = NULL;
  buf->channel_buffers = NULL;

  // Allocate a single contiguous block of memory for all channels to improve
  // cache locality and reduce memory fragmentation.
  size_t total = channels * capacity;
  buf->storage = (double*)calloc(total, sizeof(double));
  // Allocate the array of pointers that will point to the start of each
  // channel's buffer.
  buf->channel_buffers =
      (mutable_waveform_t*)malloc(channels * sizeof(mutable_waveform_t));

  if (!buf->storage || !buf->channel_buffers) {
    audio_buffers_free(buf);
    return NULL;
  }

  // Set up pointers to point into the contiguous block.
  for (size_t ch = 0; ch < channels; ch++) {
    buf->channel_buffers[ch] = buf->storage + (ch * capacity);
  }
  return buf;
}

/// Allocate a fresh buffer pool and copy existing waveform data into it.
audio_buffers_t* audio_buffers_copy_from(const double* const* waveforms,
                                         const size_t* channel_lengths,
                                         size_t channels) {
  if (channels == 0) return NULL;

  // Find the maximum length among all channels to determine the capacity of the
  // new buffer.
  size_t max_cap = 0;
  for (size_t ch = 0; ch < channels; ch++) {
    if (channel_lengths[ch] > max_cap) max_cap = channel_lengths[ch];
  }
  if (max_cap == 0) max_cap = 1;

  audio_buffers_t* buf = audio_buffers_create(channels, max_cap);
  if (!buf) return NULL;

  // Copy data from the potentially non-contiguous source waveforms into our
  // contiguous storage.
  for (size_t ch = 0; ch < channels; ch++) {
    size_t len = channel_lengths[ch];
    if (len > 0 && waveforms[ch]) {
      memcpy(buf->channel_buffers[ch], waveforms[ch], len * sizeof(double));
    }
  }
  return buf;
}

void audio_buffers_free(audio_buffers_t* buffers) {
  if (!buffers) return;
  if (buffers->storage) free(buffers->storage);
  if (buffers->channel_buffers) free(buffers->channel_buffers);
  free(buffers);
}
